### PR TITLE
feat: Slack Integration

### PR DESCRIPTION
**Pull request by Palanikannan1437**
## What does this PR do?

This PR adds in Slack integration to pipe in survery responses into your desired Slack Channels.

## Demo
https://github.com/formbricks/formbricks/assets/73993394/00262f16-fd0c-4a7e-9177-025b17714bb4

## How should this be tested?

- Create a slack app via https://api.slack.com/apps
- Keep token rotation in OAuth & Permission disabled or disable it
- In OAuth & Permissions, add `https://localhost:3000/api/auth/callback/slack` as your redirect URL
- Get the client_id and client_secret from the Basic Information tab and configure the `SLACK_CLIENT_ID` and `SLACK_CLIENT_SECRET` in your .env file
- Head over to the Formbricks main web app and go to the integrations page
- Click on the "Connect" button and complete the OAuth flow for slack after which you should be able to map channels to your surveys after a successful auth flow.

### Appreciated

- [ ] Updated the Formbricks Docs if changes were necessary --> TODO

### CODE REVIEW

**Comment by gupta-piyush19** [comment]
@Palanikannan1437  Thanks for the PR 🎊 🚀 
I am reviewing this feature; can you please sync it with the main branch and resolve merge conflicts?
We've recently merged multi-language support, for which a few changes might be required. Can you please test it end-to-end?
Feel free to reach out to me if you need some help or to discuss something 😊

File: apps/web/app/(auth)/auth/components/SlackButton.tsx
Comment (gupta-piyush19) [inline apps/web/app/(auth)/auth/components/SlackButton.tsx:20]: Can you please debug and fix the `callbackUrl`, in case of canceling the slack integration, app throws the error.
image

File: apps/web/app/api/slack/callback/route.ts
Comment (gupta-piyush19) [inline apps/web/app/api/slack/callback/route.ts:38]: Do you think we can remove the logs here?
Also, there can be an `error` queryParam in case of any error in integrating Slack, can we also handle that?

File: apps/web/package.json
Comment (gupta-piyush19) [inline apps/web/package.json:8]: Using experimental https for local development enforces getting the certificates and changes in the .env file(`WEBAPP_URL` & `NEXTAUTH_URL`) and changes in the redirect_url of other integrations. Can we mention in the docs about the https-only mode in case of slack integration?
what do you think? @mattinannt

File: packages/lib/authOptions.ts
Comment (gupta-piyush19) [inline packages/lib/authOptions.ts:206]: Do we need all these scope permissions for the integration?

**Review by gupta-piyush19 — CHANGES_REQUESTED** [review]
Thanks, @Palanikannan1437 for the PR. 🙂
The feature is working fine. I made a few tweaks to improve code consistency and refactoring.
Can you please check the comments? and add the necessary docs? 
We're good to go after that. 🚀

File: apps/web/package.json
Comment (mattinannt) [reply apps/web/package.json:8]: Maybe this is only relevant for the development & testing of the integration and we can remove them in the PR itself before merging. Later we can see if it maybe makes sense to by default use https mode for the dev server when the feature is not experimental any more.
I don't think we need to document this for now.

**Comment by Palanikannan1437** [comment]
> Thanks, @Palanikannan1437 for the PR. 🙂
> The feature is working fine. I made a few tweaks to improve code consistency and refactoring.
> Can you please check the comments? and add the necessary docs? 
> We're good to go after that. 🚀

Sure thing, I'll look into this over the weekend and resolve it, thanks 🙌

**Comment by gupta-piyush19** [comment]
Hey, @Palanikannan1437. Did you get a chance to work on it? Or I can take over and finish it.

**Comment by Palanikannan1437** [comment]
> Hey, @Palanikannan1437. Did you get a chance to work on it? Or I can take over and finish it. 

Hey @gupta-piyush19, sure thing that'd be awesome, running a bit tight on schedule 😅 🙌.

File: apps/web/app/(app)/environments/[environmentId]/integrations/slack/components/AddIntegrationModal.tsx
Comment (mattinannt) [reply apps/web/app/(app)/environments/[environmentId]/integrations/slack/components/AddIntegrationModal.tsx:291]: Also this component seems generic. Don't we maybe already have this in another integration or ui components?

File: apps/web/app/api/slack/callback/route.ts
Comment (mattinannt) [inline apps/web/app/api/slack/callback/route.ts:23]: Is this step necessary? We could also continue with `SLACK_CLIENT_ID`.
(Also otherwise a camel-case `clientId` would be the right choice)

**Review by mattinannt — CHANGES_REQUESTED** [review]
@gupta-piyush19 Thanks for making this PR ready 😊💪 I still found a few things in the code that needs a change and a few smaller issues with the functionality:

File: apps/web/app/api/slack/route.ts
Comment (gupta-piyush19) [reply apps/web/app/api/slack/route.ts:1]: @mattinannt Google sheets is also not in the integrations directly. We can also move that after updating the redirect_uri in the formbricks app in google console.

File: packages/types/integration/slack.ts
Comment (gupta-piyush19) [reply packages/types/integration/slack.ts:16]: yes, slack returns data in this format only. I've also extended the zod object even more to store all the token/ids that slack provides.

**Comment by gupta-piyush19** [comment]
Thanks for the review, @mattinannt 🚀 
I see a lot of minor changes related to the code consistency. I overlook most of the changes in the first review as they are similar to the other integrations, which raises the need to refactor the other integrations as well. Let's pick it up in some other ticket; I've fixed it for Slack now.

**Comment by mattinannt** [comment]
@gupta-piyush19 for the kamal setup we also now need the environment variables in these files:
- `.github/workflows/kamal-setup.yml`
- `.github/workflows/kamal-deploy.yml`
- `kamal/deploy.yml`
Can you please add the Slack environment variables here? :-)

**Comment by gupta-piyush19** [comment]
Thanks, @mattinannt, for the review.
I've added the env for kamal setup. Can you please re-review the PR?

**Review by mattinannt — APPROVED** [review]
@Palanikannan1437 @gupta-piyush19 Thanks for your great job! Works great and is ready to be merged! 🚀💪

**Comment by gupta-piyush19** [comment]
Thanks, @Palanikannan1437 for the PR and @mattinannt for the review(s). 🙌

**Comment by briannezhad** [comment]
This doesn't show up for us. we are using https:// and have setup the variables. version v1.6.1
Screenshot 2024-04-28 at 8 37 5\\6 PM

**Comment by briannezhad** [comment]
Thank you. 🙏

**Pull request merged**
